File: app/src/main/java/dev/ragnarok/fenrir/fragment/ShortedLinksFragment.java
package dev.ragnarok.fenrir.fragment;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.snackbar.BaseTransientBottomBar;
import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputEditText;

import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.ActivityFeatures;
import dev.ragnarok.fenrir.activity.ActivityUtils;
import dev.ragnarok.fenrir.adapter.ShortedLinksAdapter;
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment;
import dev.ragnarok.fenrir.listener.EndlessRecyclerOnScrollListener;
import dev.ragnarok.fenrir.listener.OnSectionResumeCallback;
import dev.ragnarok.fenrir.listener.PicassoPauseOnScrollListener;
import dev.ragnarok.fenrir.listener.TextWatcherAdapter;
import dev.ragnarok.fenrir.model.ShortLink;
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory;
import dev.ragnarok.fenrir.mvp.presenter.ShortedLinksPresenter;
import dev.ragnarok.fenrir.mvp.view.IShortedLinksView;
import dev.ragnarok.fenrir.place.Place;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.CustomToast;
import dev.ragnarok.fenrir.util.Utils;
import dev.ragnarok.fenrir.util.ViewUtils;

public class ShortedLinksFragment extends BaseMvpFragment<ShortedLinksPresenter, IShortedLinksView> implements IShortedLinksView, ShortedLinksAdapter.ClickListener {

    private TextView mEmpty;
    private TextInputEditText mLink;
    private MaterialButton do_Short;
    private MaterialButton do_Validate;

    private SwipeRefreshLayout mSwipeRefreshLayout;
    private ShortedLinksAdapter mAdapter;

    public static ShortedLinksFragment newInstance(int accountId) {
        Bundle args = new Bundle();
        args.putInt(Extra.ACCOUNT_ID, accountId);
        ShortedLinksFragment fragment = new ShortedLinksFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_shorted_links, container, false);
        ((AppCompatActivity) requireActivity()).setSupportActionBar(root.findViewById(R.id.toolbar));
        mEmpty = root.findViewById(R.id.fragment_shorted_links_empty_text);

        RecyclerView.LayoutManager manager = new LinearLayoutManager(requireActivity());
        RecyclerView recyclerView = root.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(manager);
        recyclerView.addOnScrollListener(new PicassoPauseOnScrollListener(Constants.PICASSO_TAG));
        recyclerView.addOnScrollListener(new EndlessRecyclerOnScrollListener() {
            @Override
            public void onScrollToLastElement() {
                callPresenter(ShortedLinksPresenter::fireScrollToEnd);
            }
        });

        mLink = root.findViewById(R.id.input_url);
        do_Short = root.findViewById(R.id.do_short);
        do_Validate = root.findViewById(R.id.do_validate);

        do_Short.setEnabled(false);
        do_Validate.setEnabled(false);
        mLink.addTextChangedListener(new TextWatcherAdapter() {
            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                do_Validate.setEnabled(!Utils.isEmpty(s));
                do_Short.setEnabled(!Utils.isEmpty(s));
                callPresenter(p -> p.fireInputEdit(s));
            }
        });

        do_Short.setOnClickListener(v -> callPresenter(ShortedLinksPresenter::fireShort));
        do_Validate.setOnClickListener(v -> callPresenter(ShortedLinksPresenter::fireValidate));

        mSwipeRefreshLayout = root.findViewById(R.id.refresh);
        mSwipeRefreshLayout.setOnRefreshListener(() -> callPresenter(ShortedLinksPresenter::fireRefresh));
        ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout);

        mAdapter = new ShortedLinksAdapter(Collections.emptyList(), requireActivity());
        mAdapter.setClickListener(this);

        recyclerView.setAdapter(mAdapter);

        resolveEmptyText();
        return root;
    }

    private void resolveEmptyText() {
        if (nonNull(mEmpty) && nonNull(mAdapter)) {
            mEmpty.setVisibility(mAdapter.getItemCount() == 0 ? View.VISIBLE : View.GONE);
        }
    }

    @Override
    public void displayData(List<ShortLink> links) {
        if (nonNull(mAdapter)) {
            mAdapter.setData(links);
            resolveEmptyText();
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        Settings.get().ui().notifyPlaceResumed(Place.DIALOGS);

        ActionBar actionBar = ActivityUtils.supportToolbarFor(this);
        if (actionBar != null) {
            actionBar.setTitle(R.string.short_link);
            actionBar.setSubtitle(null);
        }

        if (requireActivity() instanceof OnSectionResumeCallback) {
            ((OnSectionResumeCallback) requireActivity()).onSectionResume(AbsNavigationFragment.SECTION_ITEM_DIALOGS);
        }

        new ActivityFeatures.Builder()
                .begin()
                .setHideNavigationMenu(false)
                .setBarsColored(requireActivity(), true)
                .build()
                .apply(requireActivity());
    }

    @Override
    public void notifyDataSetChanged() {
        if (nonNull(mAdapter)) {
            mAdapter.notifyDataSetChanged();
            resolveEmptyText();
        }
    }

    @Override
    public void notifyDataAdded(int position, int count) {
        if (nonNull(mAdapter)) {
            mAdapter.notifyItemRangeInserted(position, count);
            resolveEmptyText();
        }
    }

    @Override
    public void showRefreshing(boolean refreshing) {
        if (nonNull(mSwipeRefreshLayout)) {
            mSwipeRefreshLayout.setRefreshing(refreshing);
        }
    }

    @Override
    public void updateLink(String url) {
        mLink.setText(url);
        mLink.setSelection(mLink.getText().length());
        do_Short.setEnabled(false);
        do_Validate.setEnabled(false);
    }

    @Override
    public void showLinkStatus(String status) {
        String stat = "";
        int color = Color.parseColor("#ff0000");

        switch (status) {
            case "not_banned":
                stat = getString(R.string.link_not_banned);
                color = Color.parseColor("#cc00aa00");
                break;
            case "banned":
                stat = getString(R.string.link_banned);
                color = Color.parseColor("#ccaa0000");
                break;
            case "processing":
                stat = getString(R.string.link_processing);
                color = Color.parseColor("#cc0000aa");
                break;
        }
        int text_color = Utils.isColorDark(color)
                ? Color.parseColor("#ffffff") : Color.parseColor("#000000");

        Snackbar.make(mLink, stat, BaseTransientBottomBar.LENGTH_LONG)
                .setBackgroundTint(color).setTextColor(text_color).setAnchorView(R.id.recycler_view).show();
    }

    @NonNull
    @Override
    public IPresenterFactory<ShortedLinksPresenter> getPresenterFactory(@Nullable Bundle saveInstanceState) {
        return () -> new ShortedLinksPresenter(
                requireArguments().getInt(Extra.ACCOUNT_ID),
                saveInstanceState
        );
    }

    @Override
    public void onCopy(int index, ShortLink link) {
        ClipboardManager clipboard = (ClipboardManager) requireActivity().getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("response", link.getShort_url());
        clipboard.setPrimaryClip(clip);
        CustomToast.CreateCustomToast(getContext()).showToast(R.string.copied);
    }

    @Override
    public void onDelete(int index, ShortLink link) {
        callPresenter(p -> p.fireDelete(index, link));
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/fragment/ShortedLinksFragment.kt
package dev.ragnarok.fenrir.fragment

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.button.MaterialButton
import com.google.android.material.snackbar.BaseTransientBottomBar
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputEditText
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.activity.ActivityFeatures
import dev.ragnarok.fenrir.activity.ActivityUtils.supportToolbarFor
import dev.ragnarok.fenrir.adapter.ShortedLinksAdapter
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment
import dev.ragnarok.fenrir.listener.EndlessRecyclerOnScrollListener
import dev.ragnarok.fenrir.listener.OnSectionResumeCallback
import dev.ragnarok.fenrir.listener.PicassoPauseOnScrollListener
import dev.ragnarok.fenrir.listener.TextWatcherAdapter
import dev.ragnarok.fenrir.model.ShortLink
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory
import dev.ragnarok.fenrir.mvp.presenter.ShortedLinksPresenter
import dev.ragnarok.fenrir.mvp.view.IShortedLinksView
import dev.ragnarok.fenrir.place.Place
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.CustomToast.Companion.CreateCustomToast
import dev.ragnarok.fenrir.util.Utils.isColorDark
import dev.ragnarok.fenrir.util.ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme

class ShortedLinksFragment : BaseMvpFragment<ShortedLinksPresenter, IShortedLinksView>(),
    IShortedLinksView, ShortedLinksAdapter.ClickListener {
    private var mEmpty: TextView? = null
    private var mLink: TextInputEditText? = null
    private var do_Short: MaterialButton? = null
    private var do_Validate: MaterialButton? = null
    private var mSwipeRefreshLayout: SwipeRefreshLayout? = null
    private var mAdapter: ShortedLinksAdapter? = null
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val root = inflater.inflate(R.layout.fragment_shorted_links, container, false)
        (requireActivity() as AppCompatActivity).setSupportActionBar(root.findViewById(R.id.toolbar))
        mEmpty = root.findViewById(R.id.fragment_shorted_links_empty_text)
        val manager: RecyclerView.LayoutManager = LinearLayoutManager(requireActivity())
        val recyclerView: RecyclerView = root.findViewById(R.id.recycler_view)
        recyclerView.layoutManager = manager
        recyclerView.addOnScrollListener(PicassoPauseOnScrollListener(Constants.PICASSO_TAG))
        recyclerView.addOnScrollListener(object : EndlessRecyclerOnScrollListener() {
            override fun onScrollToLastElement() {
                presenter?.fireScrollToEnd()
            }
        })
        mLink = root.findViewById(R.id.input_url)
        do_Short = root.findViewById(R.id.do_short)
        do_Validate = root.findViewById(R.id.do_validate)
        do_Short?.isEnabled = false
        do_Validate?.isEnabled = false
        mLink?.addTextChangedListener(object : TextWatcherAdapter() {
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                do_Validate?.isEnabled = !s.isNullOrEmpty()
                do_Short?.isEnabled = !s.isNullOrEmpty()
                presenter?.fireInputEdit(
                    s
                )
            }
        })
        do_Short?.setOnClickListener {
            presenter?.fireShort()
        }
        do_Validate?.setOnClickListener {
            presenter?.fireValidate()
        }
        mSwipeRefreshLayout = root.findViewById(R.id.refresh)
        mSwipeRefreshLayout?.setOnRefreshListener {
            presenter?.fireRefresh()
        }
        setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout)
        mAdapter = ShortedLinksAdapter(emptyList(), requireActivity())
        mAdapter?.setClickListener(this)
        recyclerView.adapter = mAdapter
        resolveEmptyText()
        return root
    }

    private fun resolveEmptyText() {
        if (mEmpty != null && mAdapter != null) {
            mEmpty?.visibility =
                if (mAdapter?.itemCount == 0) View.VISIBLE else View.GONE
        }
    }

    override fun displayData(links: List<ShortLink>) {
        if (mAdapter != null) {
            mAdapter?.setData(links)
            resolveEmptyText()
        }
    }

    override fun onResume() {
        super.onResume()
        Settings.get().ui().notifyPlaceResumed(Place.DIALOGS)
        val actionBar = supportToolbarFor(this)
        if (actionBar != null) {
            actionBar.setTitle(R.string.short_link)
            actionBar.subtitle = null
        }
        if (requireActivity() is OnSectionResumeCallback) {
            (requireActivity() as OnSectionResumeCallback).onSectionResume(AbsNavigationFragment.SECTION_ITEM_DIALOGS)
        }
        ActivityFeatures.Builder()
            .begin()
            .setHideNavigationMenu(false)
            .setBarsColored(requireActivity(), true)
            .build()
            .apply(requireActivity())
    }

    override fun notifyDataSetChanged() {
        if (mAdapter != null) {
            mAdapter?.notifyDataSetChanged()
            resolveEmptyText()
        }
    }

    override fun notifyDataAdded(position: Int, count: Int) {
        if (mAdapter != null) {
            mAdapter?.notifyItemRangeInserted(position, count)
            resolveEmptyText()
        }
    }

    override fun showRefreshing(refreshing: Boolean) {
        mSwipeRefreshLayout?.isRefreshing = refreshing
    }

    override fun updateLink(url: String?) {
        mLink?.setText(url)
        mLink?.setSelection(mLink?.text?.length ?: 0)
        do_Short?.isEnabled = false
        do_Validate?.isEnabled = false
    }

    override fun showLinkStatus(status: String?) {
        var stat = ""
        var color = Color.parseColor("#ff0000")
        when (status) {
            "not_banned" -> {
                stat = getString(R.string.link_not_banned)
                color = Color.parseColor("#cc00aa00")
            }
            "banned" -> {
                stat = getString(R.string.link_banned)
                color = Color.parseColor("#ccaa0000")
            }
            "processing" -> {
                stat = getString(R.string.link_processing)
                color = Color.parseColor("#cc0000aa")
            }
        }
        val text_color =
            if (isColorDark(color)) Color.parseColor("#ffffff") else Color.parseColor("#000000")
        mLink?.let {
            Snackbar.make(it, stat, BaseTransientBottomBar.LENGTH_LONG)
                .setBackgroundTint(color).setTextColor(text_color).setAnchorView(R.id.recycler_view)
                .show()
        }
    }

    override fun getPresenterFactory(saveInstanceState: Bundle?): IPresenterFactory<ShortedLinksPresenter> {
        return object : IPresenterFactory<ShortedLinksPresenter> {
            override fun create(): ShortedLinksPresenter {
                return ShortedLinksPresenter(
                    requireArguments().getInt(Extra.ACCOUNT_ID),
                    saveInstanceState
                )
            }
        }
    }

    override fun onCopy(index: Int, link: ShortLink) {
        val clipboard =
            requireActivity().getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager?
        val clip = ClipData.newPlainText("response", link.short_url)
        clipboard?.setPrimaryClip(clip)
        CreateCustomToast(context).showToast(R.string.copied)
    }

    override fun onDelete(index: Int, link: ShortLink) {
        presenter?.fireDelete(
            index,
            link
        )
    }

    companion object {
        fun newInstance(accountId: Int): ShortedLinksFragment {
            val args = Bundle()
            args.putInt(Extra.ACCOUNT_ID, accountId)
            val fragment = ShortedLinksFragment()
            fragment.arguments = args
            return fragment
        }
    }
}